File: frontend/lib/features/seals/seal_list_filters.dart
import 'seal_list_helpers.dart';
import 'seal_validation.dart';

/// Problémové filtry seznamu ucpávek (Task 3.5).
enum SealProblemFilter {
  noPhoto('no_photo', 'Bez fotky'),
  onePhoto('one_photo', '1 fotka'),
  awaitingReview('awaiting_review', 'Čeká kontrolu'),
  pendingSync('pending_sync', 'Čeká sync'),
  hasNote('has_note', 'Má poznámku'),
  missingData('missing_data', 'Nedokončené'),
  // Task 6 – praktické filtry.
  mine('mine', 'Moje'),
  statusDraft('status_draft', 'Rozpracované'),
  statusChecked('status_checked', 'Zkontrolované'),
  statusInvoiced('status_invoiced', 'Fakturované'),
  attention('attention', 'K řešení');

  const SealProblemFilter(this.apiValue, this.label);
  final String apiValue;
  final String label;

  static SealProblemFilter? fromApi(String value) {
    for (final f in SealProblemFilter.values) {
      if (f.apiValue == value) return f;
    }
    return null;
  }
}

List<String> sealFiltersToApi(Set<SealProblemFilter> filters) {
  return filters
      .where((f) => f != SealProblemFilter.pendingSync)
      .map((f) => f.apiValue)
      .toList();
}

bool sealMatchesFilters(
  Map<String, dynamic> seal, {
  required Set<SealProblemFilter> filters,
  required bool isWorker,
  String? currentUserId,
}) {
  if (filters.isEmpty) return true;

  final photoCount = seal['photoCount'] as int? ?? 0;
  final status = seal['status'] as String? ?? 'draft';
  final isSynced = seal['isSynced'] as bool? ?? true;

  for (final f in filters) {
    switch (f) {
      case SealProblemFilter.noPhoto:
        if (photoCount > 0) return false;
      case SealProblemFilter.onePhoto:
        if (photoCount != 1) return false;
      case SealProblemFilter.awaitingReview:
        if (status != 'draft') return false;
      case SealProblemFilter.pendingSync:
        if (isSynced) return false;
      case SealProblemFilter.hasNote:
        if (!sealHasNoteForList(seal, isWorker: isWorker)) return false;
      case SealProblemFilter.missingData:
        if (validateSealForChecked(_sealForValidation(seal)).isEmpty) {
          return false;
        }
      case SealProblemFilter.mine:
        if (currentUserId == null || seal['createdById'] != currentUserId) {
          return false;
        }
      case SealProblemFilter.statusDraft:
        if (status != 'draft') return false;
      case SealProblemFilter.statusChecked:
        if (status != 'checked') return false;
      case SealProblemFilter.statusInvoiced:
        if (status != 'invoiced') return false;
      case SealProblemFilter.attention:
        // „K řešení": nedokončené.
        final hasMissing =
            validateSealForChecked(_sealForValidation(seal)).isNotEmpty;
        if (!hasMissing) return false;
    }
  }
  return true;
}

Map<String, dynamic> _sealForValidation(Map<String, dynamic> seal) {
  final photoCount = seal['photoCount'] as int? ?? 0;
  return {
    'system': seal['system'] ?? '',
    'construction': seal['construction'] ?? '',
    'location': seal['location'] ?? '',
    'fireRating': seal['fireRating'] ?? '',
    'photos': List.generate(photoCount, (i) => {'id': '$i'}),
    'entries': seal['entries'] as List? ?? [],
  };
}

List<Map<String, dynamic>> applySealListFilters(
  List<Map<String, dynamic>> seals, {
  required Set<SealProblemFilter> filters,
  required bool isWorker,
  String? currentUserId,
}) {
  if (filters.isEmpty) return seals;
  return seals
      .where((s) => sealMatchesFilters(
            s,
            filters: filters,
            isWorker: isWorker,
            currentUserId: currentUserId,
          ))
      .toList();
}

File: backend/src/lib/seal-list-filters.ts
import { Prisma, UserRole } from '@prisma/client';
import { validateSealForChecked } from '../services/seal-validation.service.js';

export const SEAL_PROBLEM_FILTERS = [
  'no_photo',
  'one_photo',
  'awaiting_review',
  'returned',
  'has_note',
  'missing_data',
  // Task 6 – praktické filtry seznamu ucpávek.
  'mine', // jen ucpávky vytvořené přihlášeným uživatelem
  'status_draft',
  'status_checked',
  'status_invoiced',
  'attention', // „K řešení" = vrácené (returned) NEBO nedokončené (missing_data)
] as const;

export type SealProblemFilter = (typeof SEAL_PROBLEM_FILTERS)[number];

export function parseSealFilters(raw?: string | string[]): SealProblemFilter[] {
  const parts = Array.isArray(raw)
    ? raw.flatMap((v) => v.split(','))
    : (raw ?? '').split(',');
  const allowed = new Set<string>(SEAL_PROBLEM_FILTERS);
  return parts
    .map((p) => p.trim())
    .filter((p): p is SealProblemFilter => allowed.has(p));
}

export function buildSealFilterWhere(
  filters: SealProblemFilter[],
  role: UserRole,
  userId?: string,
): Prisma.SealWhereInput {
  const and: Prisma.SealWhereInput[] = [];
  for (const f of filters) {
    switch (f) {
      case 'no_photo':
        and.push({ photos: { none: {} } });
        break;
      case 'awaiting_review':
        and.push({ status: 'draft' });
        break;
      case 'returned':
        and.push({ reviewStatus: 'returned' });
        break;
      case 'status_draft':
        and.push({ status: 'draft' });
        break;
      case 'status_checked':
        and.push({ status: 'checked' });
        break;
      case 'status_invoiced':
        and.push({ status: 'invoiced' });
        break;
      case 'mine':
        // Bez userId nelze filtr aplikovat – raději nic nefiltrovat, než filtrovat chybně.
        if (userId) and.push({ createdById: userId });
        break;
      case 'has_note':
        if (role === UserRole.worker) {
          and.push({ internalNote: { not: null } });
        } else {
          and.push({
            OR: [
              { note: { not: null } },
              { internalNote: { not: null } },
            ],
          });
        }
        break;
      // 'attention' a 'missing_data' se vyhodnocují post-query (viz applyPostSealFilters).
      default:
        break;
    }
  }
  if (and.length === 0) return {};
  return { AND: and };
}

function rowPhotoCount(row: Record<string, unknown>): number {
  const count = row._count as { photos?: number } | undefined;
  return count?.photos ?? 0;
}

function rowValidationShape(row: Record<string, unknown>) {
  return {
    system: String(row.system ?? ''),
    construction: String(row.construction ?? ''),
    location: String(row.location ?? ''),
    fireRating: String(row.fireRating ?? ''),
    entries: (row.entries as Array<{
      entryType: string;
      dimension: string;
      quantity: unknown;
      materials: Array<{ material: string }>;
    }>) ?? [],
    photos: Array.from({ length: rowPhotoCount(row) }, (_, i) => ({ id: String(i) })),
  };
}

function hasMissingData(r: Record<string, unknown>): boolean {
  return validateSealForChecked(rowValidationShape(r)).length > 0;
}

export function applyPostSealFilters<T>(rows: T[], filters: SealProblemFilter[]): T[] {
  if (filters.length === 0) return rows;
  const needsPost = filters.some(
    (f) => f === 'one_photo' || f === 'missing_data' || f === 'attention',
  );
  if (!needsPost) return rows;

  return rows.filter((row) => {
    const r = row as Record<string, unknown>;
    for (const f of filters) {
      if (f === 'one_photo' && rowPhotoCount(r) !== 1) return false;
      if (f === 'missing_data' && !hasMissingData(r)) return false;
      // „K řešení": vrácené k opravě NEBO nedokončené (chybí data ke kontrole).
      if (f === 'attention' && r.reviewStatus !== 'returned' && !hasMissingData(r)) {
        return false;
      }
    }
    return true;
  });
}

export function needsEntryInclude(filters: SealProblemFilter[]): boolean {
  return filters.includes('missing_data') || filters.includes('attention');
}
